enhancements/compound-bundle-constraints.md

```diff
@@ -0,0 +1,409 @@
+- `foo.v1.0.0` cannot use a particular API version, but can use all others?
+
+These constraints are logical disjunctions and negations in character; the existing
+top-level constraints are together a logical conjunction, although not explicit.
+Currently the spec does not support specifying such "compound" dependency constraints.
+
+### Goals
+
+- Define logical conjunction, disjunction, and negation properties as dependency constraints, termed "compound constraints".
+- Define nesting of compound constraints.
+- Describe how compound constraints are evaluated.
+- Describe what needs to change in OLM's resolver to support compound constraints.
+- Plan for implementation under an alpha feature gate, and a timeline for GA.
+
+### Non-Goals
+
+- Change the existing constraint definitions.
+- Break bundle spec or OLM backwards-compatibility in any way.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I should be able to specify "any of these APIs"
+or "none of these APIs" dependencies as bundle properties, and "all of these APIs" explicitly.
+
+#### Story 2
+
+As an operator author, I should be able to nest any compound constraint(s)
+within a parent compound constraint.
+
+#### Story 3
+
+As an operator author, I should be able to build operator bundles that "work"
+with both older and newer versions of OLM. I understand that "work" might mean
+different things depending on OLM version; obviously older OLM versions cannot
+compound constraints using any method other than conjunction.
+
+#### Story 4
+
+As a cluster administrator using OLM, I should receive the same information
+about successful and failed resolution as with current "simple" constraints.
+
+#### Story 5
+
+As a cluster administrator, I should be able to opt-in to resolving compound
+constraints in OLM while the feature is alpha. When opted-out, the resolver
+should emit errors so bundles are not installed incorrectly, ex. missing an API.
+
+### Implementation Details/Notes/Constraints
+
+**Note**: see ["Implementation History"](#implementation-history) for proof-of-concept info.
+
+I propose the new [`olm.constraint` property][arb-constraints] fields `all`, `any`, and `none`
+to specify conjunction, disjunction, and negation compound constraints, respectively.
+Their value type is a struct containing a list of constraints such as `olm.package.required`,
+`olm.gvk.required`, or any compound constraint listed above under a `constraints` key.
+
+I also propose that `olm.gvk.required` and `olm.package.required` are redefined as
+`gvk` and `package` fields to align with other properties under `olm.constraint`.
+The only other difference between existing and new GVK/package constraint schemas
+is the new schemas cannot be specified as top-level, non-compound fields; they must
+always be housed under one of `all`, `any`, or `none`. Given this difference,
+OLM will only interpret one set of schemas or the other as inputs to the resolver:
```

Comment:
Shouldn't it also be possible to have something like this:

```yaml
type: olm.constraint
value:
  gvk:
    group: example.com
    version: v1
    kind: MyApp
```

Basically you can only use `gvk` and `package` within an `olm.constraint` value. But I don't think it should be required that these two types are nested within a compound constraint.

Reply:
My thought process was that top-level constraints are always a conjunction, so why not make it explicit as `all`. I am ok relaxing this given most users will probably not need compound constraints.